### PR TITLE
[language] separate CompiledScript and CompiledModule in the file format

### PR DESCRIPTION
**Pull request by vgao1996**
## Summary
This is the beginning of the work to fully separate scripts and modules. While scripts and modules share some common tables for external dependencies, the entry point (main) of the script is still very different from a full fledged function definition in a module. Therefore we thought it would be better for them to have separate views. 

This PR replaces the full function definition in CompiledScriptMut with only the fields required, also eliminating the need for a few other table entries (function handle, module handle, function name, module name, address).

Other components (verifier, vm etc.) still depend on converting scripts into modules, so until they are rewritten to handle scripts natively, we'll have keep this working. However due to the above information being removed, we'll have to do it in a slightly more complex way. See the code for details.

## Test Plan
cargo test

## Progress Tracker
- [x] File Format
- [x] IR Compiler
- [x] Move Compiler
- [x] VM
- [x] IR Tests
- [x] Move Tests
- [x] Other Tests
- [x] Documentation
- [x] Cleanup

### CODE REVIEW

File: language/compiler/ir-to-bytecode/src/compiler.rs [hunk 1]
```diff
@@ -749,6 +752,32 @@ fn compile_functions(
...
+                locals,
+                code,
+            )?
+        }
+
+        FunctionBody::Native => {
+            for (var, _) in ast_function.signature.formals.into_iter() {
+                record_src_loc!(local: context, var)
+            }
+            CodeUnit::default()
```
Comment (tnowacki) [inline language/compiler/ir-to-bytecode/src/compiler.rs:776]: Eeek! I really hate `default` for things like this. We should look into removing it at some point

File: language/vm/src/file_format.rs [hunk 2]
```diff
@@ -1365,23 +1368,130 @@ impl CompiledScript {
...
+    /// TODO: rewrite things that depend on this and get this removed.
+    #[deprecated(
+        note = "This function is deprecated and will be removed soon. Please do not introduce new dependencies."
+    )]
+    pub fn into_module(mut self) -> (ScriptConversionInfo, CompiledModuleMut) {
+        // Add the "<SELF>" identifier if it isn't present.
+        //
+        // Note: When adding an element to the table, in theory it is possible for the index
+        // to overflow. This will not be a problem if we get rid of the script/module conversion.
+        let (self_ident, self_ident_idx) = match self
```
Comment (tnowacki) [inline language/vm/src/file_format.rs:1413]: nit rename `self_ident` to `had_self_ident`?

File: language/vm/src/file_format.rs [hunk 3]
```diff
@@ -1365,23 +1368,130 @@ impl CompiledScript {
...
+        // the verifier may complain about not being able to find implementations for functioins.
+        //
+        // This is because the verifier always considers the first handle self, and it would be very
+        // tricky for us to fix this during conversion. (We don't want to shift the indices.)
+        //
+        // This will no longer be an issue once we remove script/module conversion so this won't be fixed.
+        let (self_module_handle, dummy_addr, self_module_handle_idx) =
+            if self.module_handles.is_empty() {
+                // Grab an arbitratry address. Create one if none exists.
+                let dummy_addr = self.address_identifiers.is_empty();
```
Comment (tnowacki) [inline language/vm/src/file_format.rs:1440]: nit `dummy_addr` to `needs_dummy_addr`

File: language/vm/src/file_format.rs [hunk 4]
```diff
@@ -1365,23 +1368,130 @@ impl CompiledScript {
...
+        //
+        // Note: this correctly handles scripts with either a proper self module handle or no module
+        // handles at all. For scripts whose first module handle is used to refer to external modules,
+        // the verifier may complain about not being able to find implementations for functioins.
+        //
+        // This is because the verifier always considers the first handle self, and it would be very
+        // tricky for us to fix this during conversion. (We don't want to shift the indices.)
+        //
+        // This will no longer be an issue once we remove script/module conversion so this won't be fixed.
+        let (self_module_handle, dummy_addr, self_module_handle_idx) =
```
Comment (tnowacki) [inline language/vm/src/file_format.rs:1437]: nit `self_module_handle` to `had _self_module_handle` 
nit `dummy_addr` to `needs_dummy_addr`

File: language/vm/src/file_format.rs [hunk 5]
```diff
@@ -1365,23 +1368,130 @@ impl CompiledScript {
...
 
+pub struct ScriptConversionInfo {
+    // If a dummy address was added.
+    dummy_addr: bool,
+    // If an empty signature was added.
+    dummy_sig: bool,
+    // If the <SELF> identifier was added.
+    self_ident: bool,
+    // If a dummy self module handle was added.
+    self_module_handle: bool,
```
Comment (tnowacki) [inline language/vm/src/file_format.rs:1386]: nit, maybe all of these boolean flag names should be in the form of predicates? 
`added_dummy_addr`
`added_dummy_sig`
etc

I find it very confusing when the name of something doesn't really line up with the type. For example I would expect `dummy_addr` to have type `AccountAddress` or `AddressIdentifierIndex`

File: language/compiler/ir-to-bytecode/src/compiler.rs (hunk 1 above)
Comment (vgao1996) [reply language/compiler/ir-to-bytecode/src/compiler.rs:776]: Yes that's the one of the two remaining items on my list. The other one is locals and parameters being overlapped.

File: language/vm/src/file_format.rs (hunk 5 above)
Comment (vgao1996) [reply language/vm/src/file_format.rs:1386]: That sounds much better! Thanks for suggesting this!

**Comment by vgao1996** [comment]
@wrwg @junkil-park could you please help look at the Move prover failure? It's possible that the new script/module conversion I implemented broke some assumptions...

**Comment by junkil-park** [comment]
@vgao1996, there was a typo in one of the Prover's test cases that cause the test failed. The fix just landed 1 hour ago. Could you please rebase on the upstream and force push? I will monitor if that error goes away.

**Comment by vgao1996** [comment]
Great then I don't have to cancel the landing... Yeah in case if this is still broken on master, let me know and I'd be glad to put up another fix.

**Comment by wrwg** [comment]
> @wrwg @junkil-park could you please help look at the Move prover failure? It's possible that the new script/module conversion I implemented broke some assumptions...

I just came out of a 2h meeting. Yes, this seem to have destroyed our new script integration which was just merged early this morning.

**Pull request closed without merging**